File: plugins/npm-specgen/specgen.js
#! /usr/bin/env node

const path = require('path');
const fs = require('fs');
const child_process = require("child_process");
const util = require('util');
const exec = util.promisify(child_process.exec)

const getOsName = () => {
    switch(process.platform) {
        case `darwin`:
        case `linux`:
            return process.platform
        case `win32`:
            return `windows`
        default:
            throw Error(`Unsupported platform: ${process.platform}`)
    }
}

const getExeName = (toolName) => {
    if (getOsName() === "windows") {
        return `${toolName}.exe`
    }
    return toolName
}

const getArch = () => {
    if (process.arch === "arm64")
        return "arm64"
    if (process.arch === "x64")
        return "amd64"
    throw Error(`Unsupported architecture: ${process.arch}`)
}

const getSpecgenPath = async () => {
    const osname = getOsName()
    const arch = getArch()

    const specgenPath = path.join(__dirname, `/dist/${osname}_${arch}/${getExeName("specgen")}`)
    if (!fs.existsSync(specgenPath)) {
        throw Error(`Can't find specgen tool at ${specgenPath}`)
    }
    if (osname !== `windows`) {
        const chmodCommandLine = `chmod +x ${specgenPath}`
        console.log(`Giving permissions to specgen tool: ${chmodCommandLine}`)
        try {
            const {stdout, stderr} = await exec(chmodCommandLine)
            console.error(stderr);
            console.log(stdout);
        } catch (error) {
            console.error(error.stderr)
            console.log(error.stdout)
            console.error(`Failed to grant execution permission to the specgen tool, exit code: ${error.code}`);
            console.error(error.message);
            if (error.code !== 0) { process.exit(error.code) }
        }
    }
    return specgenPath
}

const runSpecgen = async (specgenCommand) => {
    const specgenCommandLine = specgenCommand.join(" ")
    console.log(`Running specgen tool: ${specgenCommandLine}`)
    try {
        const {stdout, stderr} = await exec(specgenCommandLine)
        console.error(stderr)
        console.log(stdout)
    } catch (error) {
        console.error(error.stderr)
        console.log(error.stdout)
        console.error(`Specgen tool raised error, exit code: ${error.code}`);
        console.error(error.message);
        if (error.code !== 0) { process.exit(error.code) }
    }
}

const main = async () => {
    var args = process.argv.slice(2);
    const specgenPath = await getSpecgenPath()
    args.unshift(specgenPath)
    await runSpecgen(args)
}

main()

File: plugins/sbt-specgen/src/main/scala/Specgen.scala
package spec

import java.io.{ByteArrayOutputStream, PrintWriter}
import java.nio.file.{Files, Paths, StandardCopyOption}
import sys.process._

import sbt.{File, Logger}

object Specgen {
  def serviceScala(
    log: Logger,
    server: String,
    jsonlib: String,
    specFile: File,
    swaggerPath: File,
    generatePath: File,
    servicesPath: File,
  ): Seq[File] = {

    log.info(s"Running sbt-specgen service code generation plugin")

    val specToolPath: String = getSpecgenTool(log)

    val specCommand: Seq[String] = Seq(
      specToolPath,
      "service-scala",
      "--spec-file", specFile.getPath,
      "--swagger-path", swaggerPath.getPath,
      "--generate-path", generatePath.getPath,
      "--services-path", servicesPath.getPath
    )

    runSpecgen(log, specCommand)

    val generatedFiles = recursiveListFiles(generatePath)
    generatedFiles.toSeq
  }

  def modelsScala(
    log: Logger,
    jsonlib: String,
    specFile: File,
    generatePath: File,
  ): Seq[File] = {

    log.info(s"Running sbt-specgen models code generation plugin")

    val specToolPath: String = getSpecgenTool(log)

    val specCommand: Seq[String] = Seq(
      specToolPath,
      "models-scala",
      "--spec-file", specFile.getPath,
      "--generate-path", generatePath.getPath,
    )

    runSpecgen(log, specCommand)

    val generatedFiles = recursiveListFiles(generatePath)
    generatedFiles.toSeq
  }

  def clientScala(
    log: Logger,
    client: String,
    jsonlib: String,
    specFile: File,
    generatePath: File,
  ): Seq[File] = {

    log.info(s"Running sbt-specgen client code generation plugin")

    val specToolPath: String = getSpecgenTool(log)

    val specCommand: Seq[String] = Seq(
      specToolPath,
      "client-scala",
      "--spec-file", specFile.getPath,
      "--generate-path", generatePath.getPath,
    )

    runSpecgen(log, specCommand)

    val generatedFiles = recursiveListFiles(generatePath)
    generatedFiles.toSeq
  }

  def recursiveListFiles(path: File): Array[File] = {
    val current = path.listFiles
    current.filterNot(_.isDirectory) ++ current.filter(_.isDirectory).flatMap(recursiveListFiles)
  }

  def runSpecgen(log: Logger, specgenCommand: Seq[String]) = {
    log.info("Running specgen tool")
    log.info(specgenCommand.mkString(" "))

    val (status: Int, stdout: String, stderr: String) = runCommand(specgenCommand)

    log.info(stdout)
    log.error(stderr)

    if (status != 0) {
      throw new Exception(s"Failed to run specgen tool, exit code: $status")
    }
  }

  def runCommand(cmd: Seq[String]): (Int, String, String) = {
    val stdoutStream = new ByteArrayOutputStream
    val stderrStream = new ByteArrayOutputStream
    val stdoutWriter = new PrintWriter(stdoutStream)
    val stderrWriter = new PrintWriter(stderrStream)
    val exitValue = cmd.!(ProcessLogger(stdoutWriter.println, stderrWriter.println))
    stdoutWriter.close()
    stderrWriter.close()
    (exitValue, stdoutStream.toString, stderrStream.toString)
  }

  def getSpecgenTool(log: Logger): String = {
    val osname = getOsName()
    val arch = getArch()

    val specgenRelativePath = s"/dist/${osname}_${arch}/${getExeName("specgen")}"

    val specToolStream = getClass.getResourceAsStream(specgenRelativePath)
    val jarPath = Specgen.getClass.getProtectionDomain.getCodeSource.getLocation.getPath
    val specgenPath = jarPath.substring(0, jarPath.lastIndexOf('.'))+specgenRelativePath

    val specgenFile = new File(specgenPath)
    if (!specgenFile.exists()) {
      log.info(s"Unpacking specgen tool into: ${specgenFile.getPath}")
      val specPathParent = specgenFile.getParentFile
      if (!specPathParent.exists()) {
        specPathParent.mkdirs()
      }
      Files.copy(specToolStream, Paths.get(specgenFile.getPath), StandardCopyOption.REPLACE_EXISTING)
      specgenFile.setExecutable(true)
    }

    specgenFile.getPath
  }

  def getExeName(toolName: String): String =
    if (getOsName() == "windows")
      s"$toolName.exe"
    else
      toolName

  def getOsName(): String = {
    val osName = System.getProperty("os.name").toLowerCase()

    osName match {
      case x if x.contains("win") => "windows"
      case x if x.contains("mac") => "darwin"
      case x if x.contains("nix") || x.contains("nux") => "linux"
    }
  }

  def getArch(): String = {
    val arch = System.getProperty("os.arch")

    arch match {
      case x if x.contains("64") => "amd64"
      case _ => "x86"
    }
  }
}